### PR TITLE
Implement tree closure

### LINKED ISSUE #85: Add a closed tree state
It should be possible to prevent any further PRs from merging unless they are explicitly marked as being allowed to merge when the tree is closed.

### LINKED ISSUE #86: Latest homu not accepting r+ shorthands
I was testing it locally and it seems like homu no longer accepts `@bors-servo r+`. This is because r+ is a two-step process -- homu will leave a comment saying "commit `<sha>` approved", which contains a fully-qualified `@bors-servo r=username SHA` in an HTML comment. It then parses _that_, and approves the PR.

This was broken in https://github.com/servo/homu/pull/77/files#diff-9f3f3e773e33ae0ef2bc051f79e89e46R283 cc @cgwalters

### PR DESCRIPTION
**Pull request by Manishearth**
fixes #85 (also drive-by fixes #86)

This is implemented as `@bors-servo treeclosed=<number>`, which closes the tree for PRs below a priority value. `@bors-servo treeclosed-` (or `@bors-servo treeclosed=-1`) restores the tree.

cc @jdm @cgwalters

r? @edunham

---
This change is [Reviewable](https://reviewable.io/reviews/servo/homu/87)

### CODE REVIEW

**Comment by Manishearth** [comment]
This or at least https://github.com/servo/homu/pull/87/commits/1f2399f991d9f9d8724b9f4cc0730fb31c51c275 must land before we update the server.

--------

For testing, I set up a folder on the server `homu-test`, and a copied config with all the crates replaced with a single entry for https://github.com/manishearth/test_repo (which was set to send webhooks to build.s.o/homu-test). I also changed the port, and used `pip install -e .` so that I could hack on it directly without needing to clear the venv each time.

Sadly it can't be tested locally since homu needs to be on a public-facing server to pick up webhooks

File: homu/main.py [hunk 1]
```diff
@@ -274,7 +274,11 @@ class AuthState(IntEnum):
     NONE = 1
 
 
-def verify_auth(username, repo_cfg, state, auth, realtime):
+def verify_auth(username, repo_cfg, state, auth, realtime, my_username):
+    # homu does talk to itself in comments
```
Comment (cgwalters) [inline homu/main.py]: So I was looking at an example and was *really* confused as to how this was happening...I deduced it must be through the 📌 approval message, but it took me a while to see:

```
state.add_comment(':pushpin: Commit {:.7} has been approved by `{}`\n\n<!-- @{} r={} {} -->'.format(state.head_sha, approver, my_username, approver, state.head_sha))
```

This hidden comment recursing into the code above again...it's like recursively evil.

Can you just tweak your (python) comment to clarify that the comment it's talking about is a markdown comment in the github comment, something like:

```
# In some cases (e.g. r+) we recursively talk to ourself via a hidden markdown comment in the message
```

**Comment by cgwalters** [comment]
(For testing Homu, perhaps our OpenShift v3/Kubernetes/Docker-ized version is useful: https://github.com/cgwalters/projectatomic-ci-infra )

**Comment by cgwalters** [comment]
I only glanced at the treeclosure code...a bit lukewarm.   Feels kind of project-specific.  Maybe one way to do things like this would be to have a generic callout to another service for custom commands/hooks.  Then Homu stays as a kind of core "github PR queue engine", and one could do more complex workflows in a custom service.

That said standing up *another* service that's hooked into Homu is adding more fragility.

**Comment by Manishearth** [comment]
Updated. I don't really see it as project specific -- it is something I've wanted for both rust and servo at times and it is a pretty generic use case: when there is a PR that _must land_ before any others (e.g. fixing broken CI, rollup PRs, etc). I think that should be useful for many. It's not a feature you would use often, but it is a feature that most large projects may end up using.

**Pull request merged**